### ISSUE TITLE
New feature requested: vmap for torch.histc

### ISSUE THREAD

**Issue by rsanchezgarc** [pytorch/pytorch#99719]

**Comment by vadimkantorov** [comment]
> I am also intested in this to be able to get histograms accross a given dimension.

me too! or even to support multidim (e.g. to compute histogram of an image for each color channel for each element in the batch)

**Comment by zou3519** [comment]
Lifting the priority, but this might be difficult for us to actually do

**Comment by vadimkantorov** [comment]
I'm currently working around this need (batched histogram) by allocating the result and doing scatter_add_. Any reason why this should not work?

With more improvements like https://github.com/pytorch/pytorch/issues/61819 and https://github.com/pytorch/pytorch/issues/5405#issuecomment-1663210205 it should be fine, no?

**Comment by zou3519** [comment]
> by allocating the result and doing scatter_add_. Any reason why this should not work?

If the semantics are correct, then yeah it should work. I was worried that we would need to write some brand-new CUDA kernels, since histc has a custom CUDA kernel somewhere under the hood

**Comment by vadimkantorov** [comment]
@zou3519 also, for practical usage in computer vision, some sort of masked computation is useful (segment label map / binary mask)

**Comment by wecacuee** [comment]
For anyone curious about using `scatter_add_` as a histogram function, here's an example. It won't work for irregular bin sizes.

```python
# https://github.com/pytorch/pytorch/issues/99719#issuecomment-1664135524 
def batch_histogram(data_tensor, num_classes=-1):
    """
    Computes histograms of integral values, even if in batches (as opposed to torch.histc and torch.histogram).
    Arguments:
        data_tensor: a D1 x ... x D_n torch.LongTensor
        num_classes (optional): the number of classes present in data.
                                If not provided, tensor.max() + 1 is used (an error is thrown if tensor is empty).
    Returns:
        A D1 x ... x D_{n-1} x num_classes 'result' torch.LongTensor,
        containing histograms of the last dimension D_n of tensor,
        that is, result[d_1,...,d_{n-1}, c] = number of times c appears in tensor[d_1,...,d_{n-1}].
    """
    maxd = data_tensor.max()
    nc = (maxd+1) if num_classes <= 0 else num_classes
    hist = torch.zeros((*data_tensor.shape[:-1], nc), dtype=data_tensor.dtype, device=data_tensor.device)
    ones = torch.tensor(1, dtype=hist.dtype, device=hist.device).expand(data_tensor.shape)
    hist.scatter_add_(-1, ((data_tensor * nc) // (maxd+1)).long(), ones)
    return hist
```
with the test cases in [Google colab here](https://colab.research.google.com/drive/1rwd6LHgjWGC6kINnlQrUmbrZzLQMfalq?usp=sharing)

**Comment by vadimkantorov** [comment]
@zou3519 For this to be nicer, scalar source should be supported:
- https://github.com/pytorch/pytorch/issues/5405
